extensions-contrib/scan-query/src/main/java/io/druid/query/scan/ScanQueryEngine.java

```diff
@@ -0,0 +1,242 @@
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License. You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied. See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+package io.druid.query.scan;
+
+import com.google.common.base.Function;
+import com.google.common.base.Preconditions;
+import com.google.common.collect.Iterables;
+import com.google.common.collect.Lists;
+import com.google.common.collect.Maps;
+import io.druid.granularity.QueryGranularities;
+import io.druid.java.util.common.ISE;
+import io.druid.java.util.common.guava.BaseSequence;
+import io.druid.java.util.common.guava.Sequence;
+import io.druid.java.util.common.guava.Sequences;
+import io.druid.query.ColumnSelectorPlus;
+import io.druid.query.QueryInterruptedException;
+import io.druid.query.dimension.DefaultDimensionSpec;
+import io.druid.query.dimension.DimensionSpec;
+import io.druid.query.filter.Filter;
+import io.druid.query.select.SelectQueryEngine;
+import io.druid.segment.Cursor;
+import io.druid.segment.DimensionHandlerUtils;
+import io.druid.segment.LongColumnSelector;
+import io.druid.segment.ObjectColumnSelector;
+import io.druid.segment.Segment;
+import io.druid.segment.StorageAdapter;
+import io.druid.segment.VirtualColumns;
+import io.druid.segment.column.Column;
+import io.druid.segment.filter.Filters;
+import org.joda.time.Interval;
+
+import java.util.Arrays;
+import java.util.Iterator;
+import java.util.List;
+import java.util.Map;
+import java.util.concurrent.TimeoutException;
+
+public class ScanQueryEngine
+{
+  private static final SelectQueryEngine.SelectStrategyFactory STRATEGY_FACTORY = new SelectQueryEngine.SelectStrategyFactory();
+  public Sequence<ScanResultValue> process(
+      final ScanQuery query,
+      final Segment segment,
+      final Map<String, Object> responseContext
+  )
+  {
+    if (responseContext.get("count") != null) {
+      int count = (int) responseContext.get("count");
+      if (count >= query.getLimit()) {
+        return Sequences.empty();
+      }
+    }
+    final Long timeoutAt = (long) responseContext.get("timeoutAt");
+    final long start = System.currentTimeMillis();
+    final StorageAdapter adapter = segment.asStorageAdapter();
+
+    if (adapter == null) {
+      throw new ISE(
+          "Null storage adapter found. Probably trying to issue a query against a segment being memory unmapped."
+      );
+    }
+
+    // at the point where this code is called, only one datasource should exist.
+    String dataSource = Iterables.getOnlyElement(query.getDataSource().getNames());
```

Comment:
Preconditions.checkArgument(..) instead, looks like dataSource isn't used and this line is there only for validation?